File: Shreyansh112/cocode/web/crdt.js
// crdt.js — a browser/Node port of the Go causal-tree (RGA) sequence CRDT in
// internal/crdt. Both sides run the same algorithm so the server only has to
// relay operations; each client maps ops to positions locally.
//
// An atom is a character with a unique id {seq, site} and a parent (the id it
// was inserted after). Atoms form a tree; the visible string is a pre-order DFS
// with each node's children ordered by DESCENDING id. Deletes are tombstones.
// Operations are idempotent + commutative, and out-of-order inserts are buffered
// until their parent arrives.
(function (root) {
  "use strict";

  const ROOT = { seq: 0, site: 0 };
  const key = (id) => id.seq + ":" + id.site;

  // less: ascending order over ids.
  function less(a, b) {
    if (a.seq !== b.seq) return a.seq < b.seq;
    return a.site < b.site;
  }

  class Doc {
    constructor(site) {
      this.site = site >>> 0;
      this.clock = 0;
      this.atoms = new Map();
      this.pending = new Map(); // parentKey -> [op,...]
      this.atoms.set(key(ROOT), { id: ROOT, parent: null, char: "", deleted: false, children: [] });
    }

    nextID() {
      this.clock += 1;
      return { seq: this.clock, site: this.site };
    }

    localInsert(index, ch) {
      const parent = this.parentForIndex(index);
      // Store char as an integer code point to match Go's rune JSON encoding.
      const code = typeof ch === "string" ? ch.codePointAt(0) : ch;
      const op = { type: "insert", id: this.nextID(), parent, char: code };
      this.apply(op);
      return op;
    }

    localDelete(index) {
      const vis = this.visibleIDs();
      if (index < 0 || index >= vis.length) return null;
      const op = { type: "delete", id: vis[index] };
      this.apply(op);
      return op;
    }

    apply(op) {
      if (op.type === "insert") this._applyInsert(op);
      else if (op.type === "delete") {
        const a = this.atoms.get(key(op.id));
        if (a) a.deleted = true;
      }
    }

    _applyInsert(op) {
      const k = key(op.id);
      if (this.atoms.has(k)) return; // idempotent
      const pk = key(op.parent);
      const parent = this.atoms.get(pk);
      if (!parent) {
        const buf = this.pending.get(pk) || [];
        buf.push(op);
        this.pending.set(pk, buf);
        return;
      }
      if (op.id.seq > this.clock) this.clock = op.id.seq;
      this.atoms.set(k, { id: op.id, parent: op.parent, char: op.char, deleted: false, children: [] });
      this._insertDesc(parent.children, op.id);

      const waiting = this.pending.get(k);
      if (waiting) {
        this.pending.delete(k);
        for (const w of waiting) this._applyInsert(w);
      }
    }

    // insert id into children array keeping DESCENDING order.
    _insertDesc(children, id) {
      let i = 0;
      while (i < children.length && !less(children[i], id)) i++;
      children.splice(i, 0, id);
    }

    value() {
      let out = "";
      this._walk(ROOT, (a) => {
        if (!(a.id.seq === 0 && a.id.site === 0) && !a.deleted) {
          out += String.fromCodePoint(a.char);
        }
      });
      return out;
    }

    visibleIDs() {
      const ids = [];
      this._walk(ROOT, (a) => {
        if (!(a.id.seq === 0 && a.id.site === 0) && !a.deleted) ids.push(a.id);
      });
      return ids;
    }

    _walk(id, visit) {
      const a = this.atoms.get(key(id));
      if (!a) return;
      visit(a);
      for (const c of a.children) this._walk(c, visit);
    }

    parentForIndex(index) {
      if (index <= 0) return ROOT;
      const vis = this.visibleIDs();
      if (index - 1 < vis.length) return vis[index - 1];
      if (vis.length > 0) return vis[vis.length - 1];
      return ROOT;
    }
  }

  const api = { Doc, ROOT };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else root.CRDT = api;
})(typeof self !== "undefined" ? self : this);

File: Shreyansh112/cocode/internal/crdt/crdt.go
// Package crdt implements a sequence CRDT for collaborative text editing using
// a causal tree (a.k.a. timestamped insertion tree), which is equivalent to
// RGA (Replicated Growable Array).
//
// The idea: every inserted character is an "atom" with a globally-unique ID and
// a parent — the ID of the atom it was inserted immediately after. All atoms
// form a tree rooted at a virtual root. The visible document is the pre-order
// DFS of that tree, where the children of each node are ordered by DESCENDING
// ID. Deletions are tombstones (the atom stays in the tree, marked deleted).
//
// Why this converges: applying an atom only attaches it to its parent, and the
// sibling ordering is a pure function of IDs, so every replica builds the exact
// same tree regardless of the order in which it receives operations. The DFS is
// deterministic, therefore every replica renders the same string. Operations are
// also idempotent (re-applying an atom is a no-op) and commutative.
package crdt

import (
	"sort"
	"strings"
)

// ID uniquely identifies an atom. The zero value is the virtual Root. Ordering
// is Lamport-like: higher Seq wins, ties broken by Site, so "more recent" edits
// sort first among concurrent siblings.
type ID struct {
	Seq  uint64 `json:"seq"`
	Site uint64 `json:"site"`
}

// Root is the virtual parent of the first characters in the document.
var Root = ID{}

// less reports whether a sorts before b in ascending order.
func (a ID) less(b ID) bool {
	if a.Seq != b.Seq {
		return a.Seq < b.Seq
	}
	return a.Site < b.Site
}

// OpType is the kind of an operation.
type OpType string

const (
	OpInsert OpType = "insert"
	OpDelete OpType = "delete"
)

// Op is a replicated operation. Inserts carry Parent and Char; deletes carry
// only the target ID. Ops are the unit exchanged between replicas.
type Op struct {
	Type   OpType `json:"type"`
	ID     ID     `json:"id"`
	Parent ID     `json:"parent,omitempty"`
	Char   rune   `json:"char,omitempty"`
}

type atom struct {
	id       ID
	parent   ID
	char     rune
	deleted  bool
	children []ID // kept sorted DESCENDING by ID
}

// Doc is a single replica of a collaborative document.
type Doc struct {
	site    uint64
	clock   uint64
	atoms   map[ID]*atom
	pending map[ID][]Op // inserts waiting for an unknown parent (out-of-order)
}

// New creates an empty document replica identified by a unique site id.
func New(site uint64) *Doc {
	d := &Doc{
		site:    site,
		atoms:   map[ID]*atom{},
		pending: map[ID][]Op{},
	}
	d.atoms[Root] = &atom{id: Root}
	return d
}

// nextID returns a fresh, unique, Lamport-incremented id for a local op.
func (d *Doc) nextID() ID {
	d.clock++
	return ID{Seq: d.clock, Site: d.site}
}

// LocalInsert inserts ch at visible index (0..len) and returns the op to
// broadcast. Index 0 inserts at the very start; index == length appends.
func (d *Doc) LocalInsert(index int, ch rune) Op {
	parent := d.parentForIndex(index)
	op := Op{Type: OpInsert, ID: d.nextID(), Parent: parent, Char: ch}
	d.Apply(op)
	return op
}

// LocalDelete tombstones the atom at visible index and returns the op. The
// second result is false if the index is out of range.
func (d *Doc) LocalDelete(index int) (Op, bool) {
	vis := d.visibleIDs()
	if index < 0 || index >= len(vis) {
		return Op{}, false
	}
	op := Op{Type: OpDelete, ID: vis[index]}
	d.Apply(op)
	return op, true
}

// Apply integrates a (local or remote) op. It is idempotent and commutative.
// Inserts whose parent is not yet known are buffered and flushed once the
// parent arrives, so out-of-order delivery is handled safely.
func (d *Doc) Apply(op Op) {
	switch op.Type {
	case OpInsert:
		d.applyInsert(op)
	case OpDelete:
		if a := d.atoms[op.ID]; a != nil {
			a.deleted = true
		}
	}
}

func (d *Doc) applyInsert(op Op) {
	if _, exists := d.atoms[op.ID]; exists {
		return // idempotent
	}
	parent := d.atoms[op.Parent]
	if parent == nil {
		// Parent unknown yet: buffer until it arrives (out-of-order delivery).
		d.pending[op.Parent] = append(d.pending[op.Parent], op)
		return
	}
	// Keep the local Lamport clock ahead of anything we have seen.
	if op.ID.Seq > d.clock {
		d.clock = op.ID.Seq
	}
	a := &atom{id: op.ID, parent: op.Parent, char: op.Char}
	d.atoms[op.ID] = a
	parent.children = insertDesc(parent.children, op.ID)

	// Flush any operations that were waiting on this atom as their parent.
	if waiting, ok := d.pending[op.ID]; ok {
		delete(d.pending, op.ID)
		for _, w := range waiting {
			d.applyInsert(w)
		}
	}
}

// insertDesc inserts id into a slice kept in descending order.
func insertDesc(ids []ID, id ID) []ID {
	i := sort.Search(len(ids), func(i int) bool { return ids[i].less(id) })
	ids = append(ids, ID{})
	copy(ids[i+1:], ids[i:])
	ids[i] = id
	return ids
}

// Value renders the current visible document string.
func (d *Doc) Value() string {
	var sb strings.Builder
	d.walk(Root, func(a *atom) {
		if a.id != Root && !a.deleted {
			sb.WriteRune(a.char)
		}
	})
	return sb.String()
}

// visibleIDs returns the IDs of visible (non-deleted) atoms in document order.
func (d *Doc) visibleIDs() []ID {
	var ids []ID
	d.walk(Root, func(a *atom) {
		if a.id != Root && !a.deleted {
			ids = append(ids, a.id)
		}
	})
	return ids
}

// walk performs a pre-order DFS visiting each atom, children in descending ID.
func (d *Doc) walk(id ID, visit func(*atom)) {
	a := d.atoms[id]
	if a == nil {
		return
	}
	visit(a)
	for _, c := range a.children {
		d.walk(c, visit)
	}
}

// parentForIndex maps a visible insert index to the id to insert after.
func (d *Doc) parentForIndex(index int) ID {
	if index <= 0 {
		return Root
	}
	vis := d.visibleIDs()
	if index-1 < len(vis) {
		return vis[index-1]
	}
	if len(vis) > 0 {
		return vis[len(vis)-1]
	}
	return Root
}

// Len returns the number of visible characters.
func (d *Doc) Len() int { return len(d.visibleIDs()) }
